include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,347 @@
+        {
+            using V = std::ranges::range_value_t<U>;
+            if constexpr (!is_arrow_base_type<V>)
+            {
+                static_assert(!std::is_same_v<V, V>, "Unsupported range value type");
+            }
+        }
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const auto create_buffers = [&values]()
+        {
+            using T = std::ranges::range_value_t<ValueRange>;
+            const auto& unwrap_value = [](const T& value)
+            {
+                if constexpr (mpl::is_reference_wrapper_v<T>)
+                {
+                    return value.get();
+                }
+                else
+                {
+                    return value;
+                }
+            };
+
+            std::vector<array_data::buffer_type> buffers(2);
+            buffers[0].resize(sizeof(std::int64_t) * (values.size() + 1), 0);
+            buffers[1].resize(std::accumulate(
+                values.begin(),
+                values.end(),
+                size_t(0),
+                [&unwrap_value](std::size_t res, const auto& s)
+                {
+                    return res + unwrap_value(s).size();
+                }
+            ));
+            auto iter = buffers[1].begin();
+            const auto offsets = buffers[0].data<std::int64_t>();
+
+            size_t i = 0;
+            for (const auto& value : values)
+            {
+                const auto& unwraped_value = unwrap_value(value);
+                SPARROW_ASSERT_TRUE(
+                    std::cmp_less(unwraped_value.size(), std::numeric_limits<std::int64_t>::max())
+                );
+                offsets[i + 1] = offsets[i] + static_cast<std::int64_t>(unwraped_value.size());
+                std::ranges::copy(unwraped_value, iter);
+                std::advance(iter, unwraped_value.size());
+                ++i;
+            }
+            return buffers;
+        };
+
+        using T = std::unwrap_ref_decay_t<std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>>;
+
+        return {
+            .type = data_descriptor(arrow_type_id<T>()),
+            .length = static_cast<array_data::length_type>(values.size()),
+            .offset = offset,
+            .bitmap = bitmap,
+            .buffers = create_buffers(),
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    struct ReferenceWrapperHasher
+    {
+        template <typename T>
+        std::size_t operator()(const std::reference_wrapper<T>& ref) const
+        {
+            return std::hash<std::remove_cvref_t<T>>{}(ref.get());
+        }
+    };
+
+    struct ReferenceWrapperEqual
+    {
+        template <typename T>
+        bool operator()(const std::reference_wrapper<T>& lhs, const std::reference_wrapper<T>& rhs) const
+        {
+            return lhs.get() == rhs.get();
+        }
+    };
```

Comment:
Should I move this elsewhere ?

Reply:
Could you use `snake_case` for these? And yeah looks like it's utility but not sure where exactly to move it.

Reply:
Sure, and moved to a reference_wrapper_utils.hpp file